Purchases: add new payment method components

fditrapani opened the pull request:
#### Changes proposed in this Pull Request

This PR adds new generic payment method components in preparation for us adding additional payment method options for existing purchases. A couple notes about the PR:
- Adds a new config flag `purchases/new-payment-methods`, enabled on dev, which allows us to ship regularly to production behind the scenes. 
- Updated instances of "edit" to "change" to match our UI closer. 
- Created new Add and Change payment method components that are now placed in the `manage-purchase` folder because that's where these components are accessed from. These will replace the existing Add and Edit Credit card components eventually.
- Created new `PaymentMethodForm` component which is a mirror of our current `CreditCard` form component. This is where we will begin adding more payment methods. 

**Before**
![image](https://user-images.githubusercontent.com/6981253/100790394-b6a92d80-33e5-11eb-8689-8a87c4a8f8dd.png)

**After**
![image](https://user-images.githubusercontent.com/6981253/100790796-58307f00-33e6-11eb-9187-50ef1c852e60.png)


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Add `flags=purchases/new-payment-methods` query string if you're not testing locally
* Visit purchases screens at the site and account level
* Start by visiting the payment method tab and click "Add payment method" button. Confirm that you see the credit card form and the page title says "Add Payment Method". 
* Sandbox to confirm that you can actually add the card.
* Visit the manage purchase screen for an existing purchase that doesn't have a payment method.
* Click on "Add payment method" and confirm that you see the credit card form and the page title says "Add Payment Method".
* Sandbox to confirm that you can actually add the card.
* Visit the manage purchase screen for an existing purchase that does have a payment method attached. 
* Click on "Change payment method" and confirm that you see the credit card form and the page title says "Change Payment Method". 
* Sandbox to confirm that you can actually change the card.

Fixes https://github.com/Automattic/wp-calypso/issues/47929

Commit:
Create new payment method components

Commit:
Create new payment method components

Commit:
Add routes

Commit:
Add routes

Commit:
Add config flag

Commit:
Add config flag

Commit:
Add logic with config flag

Commit:
Add logic with config flag

Commit:
Update edit > change to reflect UI language

Commit:
Update edit > change to reflect UI language

Commit:
Update style names

Commit:
Update style names

Commit:
Update more instances of 'edit' to 'change'

Commit:
Update more instances of 'edit' to 'change'

donlair committed:
Adding WP Courses task to My Home. (#47721)

* Adding WP Courses task to My Home.

* Adding Google Analytics tracking params to link.

* Updating card copy.

* Update client/my-sites/customer-home/cards/tasks/wp-courses/index.jsx

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

sirbrillig commented on client/me/purchases/manage-purchase/change-payment-method/index.jsx:
```diff
@@ -0,0 +1,144 @@
+import PaymentMethodSidebar from 'calypso/me/purchases/components/payment-method-sidebar';
+import PaymentMethodLoader from 'calypso/me/purchases/components/payment-method-loader';
+
+function EditPaymentMethod( props ) {
```
This component should probably be renamed to match the filename.

fditrapani replied:
Thanks for catching that. I've updated it.

sirbrillig reviewed (commented):
I followed the testing instructions and this works as described... but I wonder if I'm testing the right thing? Here's what I'm confused about:

If I visit `/purchases/payment-methods/example.com?flags=purchases/new-payment-methods` and then click "Add payment method", I'm redirected to `/purchases/add-credit-card/example.com`, not the new `add-payment-method` route.

Oh! I just realized that this sets the new flag to `false` for wpcalypso, and I was testing using calypso.live, which uses that config. Could we also set it to true there?

p-jackson committed:
Onboarding: Add support for RTL site launch stylesheet to ETK (#47890)

p-jackson committed:
Onboarding: Add RTL support to domain picker (#47889)

sirbrillig committed:
Shopping cart: Fix cart key change and reload behavior (#47625)

* Add cart prop to withShoppingCart

* Make all args to getCartKey optional

* Update ShoppingCartProvider type so it can accept multiple children

* Change CART_RELOAD action to not erase cart

* Move cartKey change check to useShoppingCartManager

The previous location for this logic was blocked unless the cacheStatus
is fresh.

* Ignore CART_RELOAD when queuing actions in shoppingCartReducer

* Strip out credits products when they come back from the server

* Remove code that strips credit products from calypso

* Restore erasing cart during reload

* Remove credits line item removal test from calypso

Calypso should now never see one of these line items.

* Add emptyResponseCart to shopping-cart package API

* Make getEmptyResponseCart a function to prevent mutation

Commit:
Update component name and turn on wpcalypso flag

Commit:
Update component name and turn on wpcalypso flag

gibrown committed:
Jetpack Search: Cleanup and remove flag for enabling it since we are fully launched. (#47479)

fditrapani commented:
> Oh! I just realized that this sets the new flag to false for wpcalypso, and I was testing using calypso.live, which uses that config. Could we also set it to true there?

Sure thing. New commit incoming!

bluefuton committed:
Reader: fix follow count for Tracks events (#47818)

* Docs fixes

* Set up a Reader analytics middleware

* Add follow count in middleware

* Use thunk rather than middleware

gibrown committed:
Jetpack Search: switch upsell to default to monthly (#47481)

gibrown committed:
Reader: remove Discover since no longer getting updated (#47880)

Commit:
Update button for expired credit cards

Commit:
Update button for expired credit cards

sirbrillig commented:
> Oh! I just realized that this sets the new flag to false for wpcalypso, and I was testing using calypso.live, which uses that config. Could we also set it to true there?

I'm sorry, I thought about this some more and I'm wrong; the whole point of using `flags=` is because it wasn't set by the config 🤦. I don't know why I asked for that to be changed. Anyway, no harm done, but sorry to waste your time on that part.

However, my question remains. When I test this and click on "Add payment method", I'm being redirected to `add-credit-card/` and not `add-payment-method/`. Is that expected?

sirbrillig committed:
Update PopoverCart to use shopping-cart package (#47605)

* Add ShoppingCartProvider to DomainList test

* Update PopoverCart to use withShoppingCart

* Update HeaderCart to use withShoppingCart

* Update CartToggleButton to be a component in PlansNavigation

* Remove unused cart prop from PlansNavigation

* Remove CartData from instances of PlansNavigation

* Make it clear that cart is required by HeaderCart

* Remove closeSectionNavMobilePanel from PopoverCart

* Add closeSectionNavMobilePanel to CartToggleButton in PlansNavigation

* Simplify getCartKey call in EditorCheckoutModal

This just removes some unnecessary values and one TypeScript error
having to do with `undefined` vs. `null`.

* Add CalypsoShoppingCartProvider

* Use CalypsoShoppingCartProvider in EditorCheckoutModal

* Use CalypsoShoppingCartProvider in CheckoutSystemDecider

* Use CalypsoShoppingCartProvider to wrap DomainSearch

* Use CalypsoShoppingCartProvider in PlansNavigation

* Replace reloadCart with reloadFromServer in HeaderCart

* Replace reloadCart with reloadFromServer in PopoverCart

* Change CartItem to use withShoppingCart for deletion

* Add CalypsoShoppingCartProvider to DomainManagementData

bluefuton committed:
Reader Lists: use Intl.Collator to offer locale-aware sorting of sidebar (#47568)

* Reader Lists: use Intl.Collator to offer locale-aware sorting of sidebar

* Switch to use selector

* Adjust naming to reflect that we're using current locale, not current user locale

* Handle missing state.ui

fditrapani commented:
> Anyway, no harm done, but sorry to waste your time on that part.

No worries! I think it's better this way because then it's easier for people to test. 

> However, my question remains. When I test this and click on "Add payment method", I'm being redirected to add-credit-card/ and not add-payment-method/. Is that expected?

oh! I see what you mean. I updated the routes at the account level and then swapped the components differently at the site level. I'll get that cleaned up.

Commit:
Fix add payment method url

Commit:
Fix add payment method url

fditrapani commented:
All fixed up @sirbrillig. Let me know how that looks now.

ramonjd committed:
Adding an is404 check to capture the case where the user edits the post's visibility in the editor, by switching the post status to private. Individual private posts are not available at all to the public and will 404. Here we're letting 404s through so we can check if coming soon is enabled. (#47945)

p-jackson committed:
Onboarding: Flip the domain picker illustration in RTL languages (#47891)

lsl committed:
Use LocaleProvider in calypso (#47613)

Co-authored-by: Philip Jackson <p-jackson@live.com>

tyxla committed:
Plugins: Fix mass removal (#47949)

tyxla committed:
Plugins: Remove old autoupdate flux action types (#47902)

tyxla committed:
Plugins: Reduxify removePluginUpdateInfo flux action (#47908)

* State: Dispatch plugin update info flux action from Redux

* Plugins: Remove usage of removePluginUpdateInfo flux action

* Plugins: Remove removePluginUpdateInfo flux action

tyxla committed:
Plugins: Reduxify PluginActivateToggle progress (#47920)

tyxla committed:
Plugins: Reduxify PluginAutoUpdateToggle progress (#47921)

tyxla committed:
Plugins: Remove unnecessary NOT_ALLOWED_TO_RECEIVE_PLUGINS flux action (#47911)

scinos committed:
Initial import of webpack-rtl-plugin (#47904)

* Initial import of webpack-rtl-plugin
* Adapt webpack-rtl-plugin to our repo structure (#47906)

sgomes committed:
Modularise immediate login state (#47914)

Fixes #42445.

sgomes committed:
Modularise media state (#47917)

Fixes #42456.

scinos committed:
Security updates from Dependabot (#47892)

* Bump bl from 1.2.2 to 1.2.3 (#47632)

Bumps [bl](https://github.com/rvagg/bl) from 1.2.2 to 1.2.3.
- [Release notes](https://github.com/rvagg/bl/releases)
- [Commits](https://github.com/rvagg/bl/compare/v1.2.2...v1.2.3)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump prismjs from 1.17.1 to 1.21.0 (#47627)

Bumps [prismjs](https://github.com/PrismJS/prism) from 1.17.1 to 1.21.0.
- [Release notes](https://github.com/PrismJS/prism/releases)
- [Changelog](https://github.com/PrismJS/prism/blob/master/CHANGELOG.md)
- [Commits](https://github.com/PrismJS/prism/compare/v1.17.1...v1.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump elliptic from 6.5.2 to 6.5.3 (#47631)

Bumps [elliptic](https://github.com/indutny/elliptic) from 6.5.2 to 6.5.3.
- [Release notes](https://github.com/indutny/elliptic/releases)
- [Commits](https://github.com/indutny/elliptic/compare/v6.5.2...v6.5.3)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump websocket-extensions from 0.1.3 to 0.1.4 (#47630)

Bumps [websocket-extensions](https://github.com/faye/websocket-extensions-node) from 0.1.3 to 0.1.4.
- [Release notes](https://github.com/faye/websocket-extensions-node/releases)
- [Changelog](https://github.com/faye/websocket-extensions-node/blob/master/CHANGELOG.md)
- [Commits](https://github.com/faye/websocket-extensions-node/compare/0.1.3...0.1.4)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jsnajdr committed:
Upgrade to wp-prettier 2.2.1 (#47950)

* Upgrade wp-prettier to 2.2.1 from NPM

* Reformat with Prettier 2.2.1

* Fix JSDoc ESLint violations in post-normalizer

mmtr committed:
Use new inclusive site settings (#47954)

mreishus committed:
Rename main function inside use-on-login.ts to useOnLogin (#47933)

ChaosExAnima committed:
Stores redirect starting in state to prevent additional redirects. (#47937)

rcanepa committed:
Show new Jetpack plans benefits in the Disconnect Site page (#47961)

* Enable features for new plans

* Replace daily with real-time backups

* Real-time scanning for Security Real-time and Complete

sirbrillig committed:
Shopping cart: Add Promise to each action that resolves when cart updates (#47648)

* Return isPendingUpdate true from useShoppingCart if queued actions

* Return Promise from cart actions and resolve when cart next valid

* Mention Promise in shopping-cart README

* Allow having many CartValidCallback Promises

* Add basic test of addProductsToCart

* Add test for addProductsToCart Promise

* Add tests for removeProductFromCart

* Add test for replaceProductsInCart

* Add test for replaceProductInCart

* Add test for applyCoupon

* Add test for removeCoupon

* Add test for updateLocation

* Make explicit types for cart tax data

* Add test for reloadFromServer

stacimc committed:
Correctly pass subscribeButton to Buttons block (#47944)

Correctly passes the subscribeButton prop in to the Buttons block
so that it can sync the planId between Premium Content and its
nested Recurring Payments block.

pottedmeat commented on client/me/purchases/payment/edit-card-details/index.jsx:
```diff
@@ -92,7 +92,7 @@ function EditCardDetails( props ) {
 						fetchStripeConfiguration={ getStripeConfiguration }
 					>
-						<CreditCardForm
+						<PaymentMethodForm
```
I'm not sure I follow why the previous `EditCardDetails` component is getting the new `PaymentMethodForm` and the new `ChangePaymentMethod` is using the previous `CreditCardForm`. Could you help explain? Thanks.

sirbrillig commented on client/me/purchases/payment/edit-card-details/index.jsx:
```diff
@@ -92,7 +92,7 @@ function EditCardDetails( props ) {
 						fetchStripeConfiguration={ getStripeConfiguration }
 					>
-						<CreditCardForm
+						<PaymentMethodForm
```
I'm probably misunderstanding something, but I'm confused that here we're replacing `CreditCardForm` with `PaymentMethodForm` for everyone, in `AddNewPaymentMethod`, we're replacing it only when the flag is set, and in `ChangePaymentMethod` we're not replacing it at all. Are those choices intentional?

fditrapani replied:
Sorry about the mixup. There's a lot going here. You are both right that I should leave things as is and use the flag to display the `ChangePaymentMethod` component instead of this one.

sirbrillig commented on client/my-sites/purchases/index.js:
```diff
@@ -110,10 +110,10 @@ export default ( router ) => {
 
 	page(
-		'/purchases/add-credit-card/:site',
+		'/purchases/add-payment-method/:site',
```
Is it intentional to change this route for everyone, regardless of the flag?

fditrapani replied:
At one point I had both and then I decided to change it for everyone. In retrospect, it's probably better to keep all the new updates behind the flag and otherwise keeps things as is for everyone else.

sirbrillig commented on client/my-sites/purchases/payment-methods/index.tsx:
```diff
@@ -99,8 +100,25 @@ export function AddNewPaymentMethod( { siteSlug }: { siteSlug: string } ): JSX.E
+				path={
+					config.isEnabled( 'purchases/new-payment-methods' )
+						? '/purchases/add-payment-method'
+						: '/purchases/add-credit-card'
```
Correct me if I'm wrong, but `/purchases/add-credit-card` no longer exists, right? In this PR the route has been replaced with `/purchases/add-payment-method` (but maybe that wasn't intentional).

fditrapani replied:
Removing the old route was the mistake. I'll keep this in place and reinstate the old route.

fditrapani replied:
This should all be resolved with the latest commit.

sirbrillig reviewed (commented):
I'm a little confused about what changes in this PR are intended to be behind the flag and which ones are intended to be for everyone. It might be worth breaking the two into separate PRs to make it easier to see the difference.

pottedmeat committed:
Add ability to redirect when completing a purchase triggered by ManagePurchase (#47928)

* Move handleRenew* optional props to an object
* Add test for new redirectTo option
* Add redirect prop and pass to handler
* Pass site-level redirect to component

fditrapani commented:
> I'm a little confused about what changes in this PR are intended to be behind the flag and which ones are intended to be for everyone. It might be worth breaking the two into separate PRs to make it easier to see the difference.

My apologies @sirbrillig , things definitely got muddled up because I started making some changes for everyone and then put some of them behind the flag. I'd like to take another shot at cleaning this up so that I'm not introducing anything new for anyone that's not behind the flag. I found it tricky not applying the components updates without seeing the results via flag so I'd like to keep it all in one PR is that's ok.

Commit:
Revert old component to not use new component

Commit:
Revert old component to not use new component

p-jackson committed:
Signup: remove the unused `skippableDomainStep` AB test (#47946)

sdnunca committed:
Editing toolkit: Fixes missing block translation (#47953)

stacimc committed:
Editing Toolkit: Update to 2.8.10 (#47969)

Commit:
Add flags to routing

Commit:
Add flags to routing

noahtallen committed:
Editing Toolkit: Update wp-env version (#47979)

This fixes issues with running `npx wp-env *` commands using the local wp-env version.

Some examples of what this fixes:
- Phpunit tests in CI are working now.
- Using `npx wp-env run wordpress bash` works now.

fditrapani commented:
Alright @sirbrillig and @pottedmeat , I think I got this right now. Here's another recap of what's going on:

- I added the config flag for the routing so that the old urls are in place when the flag is disabled. 
- At the account level we have  new components to add a payment method to the account, change payment method for a purchase, and add payment method to a purchase compared to the site level where the components were all fine as is and I added conditions within them to replace the `CreditCard` form with the `PaymentMethodForm`. Bigger changes were needed at the account level because of where the components were located (in a payments folder) and they were credit card specific.

sirbrillig committed:
Remove useShowAddCouponSuccessMessage (#47943)

It is unnecessary as the shopping-cart endpoint already includes such a
message.

sirbrillig commented on client/me/purchases/components/payment-method-form/helpers.js:
```diff
@@ -0,0 +1,229 @@
+
+const wpcom = wpcomFactory.undocumented();
+
+export async function saveOrUpdateCreditCard( {
```
Since all these functions are just being duplicated without any changes (I think; correct me if I'm wrong), we could change the "old" code to pull these functions from this file instead. It might not be worth it, but I'm curious what you think.

My reasoning is that we're working on the logic around these functions in https://github.com/Automattic/wp-calypso/pull/47642 and if we have a duplicate copy here, we're going to need to make sure we make changes in both places. I'm worried that something will be fixed in the old location and not fixed here, and then we'll eventually remove the old location with all its fixes.

fditrapani replied:
> we could change the "old" code to pull these functions from this file instead.

Yeah, that makes sense to me. My thinking was that we would delete the old file once we were ready with this one.

sirbrillig commented on client/me/purchases/add-new-payment-method/index.jsx:
```diff
@@ -0,0 +1,83 @@
+import PaymentMethodSidebar from 'calypso/me/purchases/components/payment-method-sidebar';
+
+function AddNewPaymentMethod( props ) {
+	const createAddCardToken = ( ...args ) => createCardToken( 'card_add', ...args );
```
This, too, presents a risk because the original component is being changed in https://github.com/Automattic/wp-calypso/pull/47642, although it's not as obvious how to prevent the duplication in this case.

fditrapani replied:
hmm, yeah that's tricky. One thought that comes to mind is to extra, and centralize the parts of this that both the components would need. Does that make sense?

sirbrillig reviewed (commented):
Code looks good to me, and testing works as described. Thanks!

My only concern is that this duplicates quite a lot of code that is undergoing some fixes in https://github.com/Automattic/wp-calypso/pull/47642 and I'm worried that something will remain broken in the copy. My suggestion, if the time frame is reasonable, is to wait for that PR to merge, and then update this one to re-duplicate the files so that those fixes get included here.

The only other options I can think of are to make sure that PR applies the fixes to both locations, which is going to make it much harder to test, or to refactor things in this PR so that the important points are not duplicated, but that would be quite a lot more work.

fditrapani commented:
Thanks for taking a look @sirbrillig. 

> My only concern is that this duplicates quite a lot of code that is undergoing some fixes in #47642 and I'm worried that something will remain broken in the copy. My suggestion, if the time frame is reasonable, is to wait for that PR to merge, and then update this one to re-duplicate the files so that those fixes get included here.

Alternatively, if this is looking good, we can maybe merge it and then centralize the common pieces to reduce the duplicated parts?

fditrapani commented:
>  My suggestion, if the time frame is reasonable, is to wait for that PR to merge, and then update this one to re-duplicate the files so that those fixes get included here.

Another thought I had was that maybe we can consolidate the two components sooner and use the flags to manage the differences when they come. In this case, the consolidation can happen separately right after both branches merge.

scinos committed:
Keep artifacts only for trunk and failed builds (#47955)

ddc22 committed:
Fix broken mobile arrow scroll of the plans page in  safari/mobile (#47952)

Fix broken mobile arrow scroll of the plans page in safari/mobile

tyxla committed:
Plugins: Introduce and use new plugin actions status selectors (#47951)

* State: Introduce plugin action status selectors

* Plugins: Update PluginActivateToggle to use isPluginActionInProgress

* Plugins: Update PluginAutoUpdateToggle to use isPluginActionInProgress

tyxla committed:
Plugins: Cleanup unused notices prop and shouldComponentUpdate (#47964)

* Plugins: Cleanup unused notices prop and shouldComponentUpdate

* Remove compose

niranjan-uma-shankar committed:
Quick Start booking page: Show morning and evening times separately (#46999)

* WIP commit

* Hide the time group if no value exists

* Switch from segmented control to button to match the UI

* Switch from segmented control to button to match the UI

* Convert 12 AM to midnight

* Add title for the button when no slots available

* Add optgroup for each time choice

* Add optgroup for each time choice

* Add icons for the time group filter button

* A few minor fixes

* Styling changes

* 1. Remove timezone from dropdown, and show it below the dropdown 2. Code refactoring to address review feedback

* Reuse string for existing translations

jeherve committed:
Calypso editor: do not set Calypsoify on Jetpack 9.2 (#47925)

unDemian committed:
Add P2 all section (#47787)

* Add P2 all section

* Ignore double cast for now

tyxla committed:
Plugins: Reduxify plugin fetching (#47915)

* State: Move plugin receiving in a separate action creator

* State: Move site plugin fetching to a separate action creator

* State: Dispatch flux plugins fetch actions in Redux

* Plugins: Use Redux instead of flux for fetching plugins

* Plugins: Remove unused fetchSitePlugins flux action

* Plugins: Add deprecation notice for fetchSitePlugins flux action

* State: Update plugin fetching tests

* Remove unused import

sirreal committed:
Devdocs: Remove experimental WordPress components (#47907)

fditrapani committed:
Domains: fix auto-renew alignment (#47958)

* Vertically align domain auto-renew

* Remove flex business

Co-authored-by: fditrapani <filippo.di.trapani@a8c.com>

Commit:
Focused Launch: Add persistent success view after launch (#47808)

Use shouldDisplaySuccessView, enablePersistentSuccessView, and disablePersistentSuccessView to control post-launch success view being displayed over the editor.

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

Commit:
Remove `jetpackConversionRateOptimization` experiment (#47940)

rcanepa committed:
Jetpack plans: add 'Downgrade to another plan' as a new cancellation reason (#47972)

cpapazoglou committed:
Nav Unification: use new experiment for internal testing: (#47996)

that experiment cannot be accidentally run through the interface

Co-authored-by: cpap <papazoglou.charalampos@gmail.com>

fditrapani committed:
Purchases: Update billing history search (#48002)

* Purchases: Update billing history search

* Remove initial value because it's null anyways

* Fix team city test

* Remove unused selector and update test rule

Co-authored-by: fditrapani <filippo.di.trapani@a8c.com>
Co-authored-by: Michael Cain <michael@michaeldcain.com>

sirbrillig committed:
Shopping cart: Update StaleCartItemsNotice to use useShoppingCart (#47970)

* Modify StaleCartItemsNotice to use withShoppingCart

* Convert StaleCartItemsNotice to functional component

This lets use use useEffect to postpone the notice display functionality
until after render, preventing warnings that we cannot render while
other renders are in-progress.

* Wrap StaleCartItemsNotice in CalypsoShoppingCartProvider

yansern committed:
Shopping Cart: Use import type where it should. (#48005)

* Use `import type` where it should.

* Add "importsNotUsedAsValues": "error" to tsconfig.

rcanepa committed:
Jetpack Products: always add products to cart by URL (#48001)

sirbrillig committed:
Shopping cart: replace existing cart for renewals whenever any product is added (#47978)

* Modify addItemsToResponseCart to replace cart sometimes

If any item that would be added is a renewal (except for
domain_redemption for some reason), then the cart is replaced entirely.

If any item in the existing cart is a renewal and any item that would be
added is not a renewal, then the cart is replaced entirely.

This copies over some of the logic (minus the Jetpack plans logic) from
cartItemShouldReplaceCart.

* Change useAddProductsFromUrl to add renewals with other products

Previously in https://github.com/Automattic/wp-calypso/pull/46248, code
was added to useAddProductsFromUrl so that renewals would be added
separately and would replace the cart. However, the logic we want is
more complex than that (see old cartItemShouldReplaceCart function) and
it would be nice to apply it to all uses of `addProductsToCart`, so here
we remove those additions and leave it to `useShoppingCart` to handle
when to replace the cart.

* Add missing quantity to createRenewalItemToAddToCart

quantity was added as a required property to RequestCartProduct in
https://github.com/Automattic/wp-calypso/pull/47700

* Update addProductsToCart in README to mention replacement

* Add debug logs to addItemsToResponseCart

* Fix tests for cart-actions

* Allow adding multiple renewals to the cart

* Add tests for add/replace product behavior

alshakero committed:
Focused Launch: Implement step highlighting logic (#47503)

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>
Co-authored-by: Razvan Papadopol <razvan.papadopol@automattic.com>

elliottprogrammer committed:
Billing: Fix - Purchase plan details stuck in loading state. (#47982)

mreishus committed:
Refactor createSite() to take "named parameters" object (#47980)

donlair committed:
Adding placeholder link to upcoming Payments tutorial video section. (#47602)

* Adding placeholder link to upcoming Payments tutorial video section.

* Updating destination url to point to the actual support page.

* Reducing font weight of link text.

blackjackkent committed:
Updating name of feature flag for anchor.fm project (#48020)

sirbrillig committed:
Revert "Billing: Fix - Purchase plan details stuck in loading state. (#47982)" (#48022)

This reverts commit 19e70f39756d466f6cbf54a865e888906ca76a6e.

nbloomf committed:
Fix add-card page for Brazil addresses (#47642)

Simplify the CreditCardForm component and its dependencies to only use Stripe fields in all contexts, rather than sometimes using our custom country specific fields.

Co-authored-by: nbloomf <nathan.bloomfield@automattic.com>
Co-authored-by: Payton Swick <payton@foolord.com>

andrewserong committed:
Patterns: Add isset check before attempting to access index on $split_slug (#48025)

Commit:
Rebase trunk

fditrapani commented:
Apologies for the ping here everyone. Something blew up as I was rebasing.

fditrapani commented:
I'm going to close this PR and open up a new one.
